### PR TITLE
Update mermaid graph of onigumo processing with a new approach

### PR DESCRIPTION
**Pull request by nappex**

### CODE REVIEW

**Comment by nappex** [comment]
You can play with graph by live service https://mermaid.live/

**Review by Glutexo — CHANGES_REQUESTED** [review]
Changing the diagram causes a slight mismatch in the description above it: although “Onigumo is composed of three sequentially interconnected components” is still correct in some sense, it doesn’t mention the new Feeder part. I’d say something like “The crawling part of Onigumo is composed of three sequentially interconnected components”. We can do that in a follow-up pull request.

It would be possible to object that the diagram doesn’t reflect the actual architecture of the project at the moment, but it didn’t do so before the change either: there is no Parser (will be resolved by #89), no Operator, no Materializer, no flow from start to end… So I think we don’t need to wait with merging for the actual feeder to emerge.

File: README.md [hunk 1]
```diff
@@ -16,27 +16,47 @@ The flowchart below illustrates the flow of data between those parts:
...
+        spider_operator(🕷️ OPERATOR)
+        onigumo_downloader[DOWNLOADER]
     end
+
+    start([START])  --> onigumo_feeder[FEEDER]
+    onigumo_feeder  -- .raw --> Crawling
+    onigumo_feeder  -- .urls --> Crawling
+    onigumo_feeder  -- .json --> Crawling
+    Crawling  --> spider_materializer(🕷️ MATERIALIZER)
+    spider_materializer --> done([END])
```
Comment (Glutexo) [inline README.md:31]: Can we only use a single space,

```suggestion
    start([START]) --> onigumo_feeder[FEEDER]
    onigumo_feeder -- .raw --> Crawling
    onigumo_feeder -- .urls --> Crawling
    onigumo_feeder -- .json --> Crawling
    Crawling --> spider_materializer(🕷️ MATERIALIZER)
    spider_materializer --> done([END])
```

or align the items to table columns?

```suggestion
    start([START])               --> onigumo_feeder[FEEDER]
    onigumo_feeder      -- .raw  --> Crawling
    onigumo_feeder      -- .urls --> Crawling
    onigumo_feeder      -- .json --> Crawling
    Crawling                     --> spider_materializer(🕷️ MATERIALIZER)
    spider_materializer          --> done([END])
```

I don’t have preference (at least now), but I’d like it to be at least consistent.

File: README.md [hunk 2]
```diff
@@ -16,27 +16,47 @@ The flowchart below illustrates the flow of data between those parts:
...
+
+    spider_operator     -. "<hash>.urls" .-> onigumo_downloader
+    onigumo_downloader -. "<hash>.raw"  .->  spider_parser
+    spider_parser     -. "<hash>.json" .->  spider_operator
```
Comment (Glutexo) [inline README.md:35]:
```suggestion
    onigumo_downloader -. "<hash>.raw" .-> spider_parser
    spider_operator -. "<hash>.urls" .-> onigumo_downloader
    spider_parser  -. "<hash>.json" .-> spider_operator
```

```suggestion
    onigumo_downloader -. "<hash>.raw"  .-> spider_parser
    spider_operator.   -. "<hash>.urls" .-> onigumo_downloader
    spider_parser      -. "<hash>.json" .-> spider_operator
```

File: README.md [hunk 3]
```diff
@@ -16,27 +16,47 @@ The flowchart below illustrates the flow of data between those parts:
...
+
+    subgraph Onigumo
+        onigumo_feeder[FEEDER]
+        onigumo_downloader[DOWNLOADER]
+    end
+
+    onigumo_feeder  -- .json --> spider_operator
+    spider_operator ---> spider_materializer
+    onigumo_feeder  -- .urls --> onigumo_downloader
+    onigumo_feeder  -- .raw --> spider_parser
```
Comment (Glutexo) [inline README.md:55]:
```suggestion
     onigumo_feeder -- .json --> spider_operator
     spider_operator ---> spider_materializer
     onigumo_feeder -- .urls --> onigumo_downloader
     onigumo_feeder -- .raw --> spider_parser
```

```suggestion
     onigumo_feeder  -- .json  --> spider_operator
     spider_operator          ---> spider_materializer
     onigumo_feeder  -- .urls  --> onigumo_downloader
     onigumo_feeder  -- .raw   --> spider_parser
```

File: README.md [hunk 4]
```diff
@@ -16,27 +16,47 @@ The flowchart below illustrates the flow of data between those parts:
...
+
+    spider_parser     -. "<hash>.json" .->  spider_operator
+    onigumo_downloader -. "<hash>.raw"  .->  spider_parser
+    spider_operator     -. "<hash>.urls" .-> onigumo_downloader
```
Comment (Glutexo) [inline README.md:59]:
```suggestion
    onigumo_feeder  -- .urls --> onigumo_downloader
    onigumo_feeder  -- .raw --> spider_parser

    spider_parser     -. "<hash>.json" .->  spider_operator
    onigumo_downloader -. "<hash>.raw"  .->  spider_parser
    spider_operator     -. "<hash>.urls" .-> onigumo_downloader

    spider_operator ---> spider_materializer
```

Let’s push the `spider_operator ---> spider_materializer` relation to the bottom, so the Spider subgraph arrows are coupled together

**Review by Glutexo — CHANGES_REQUESTED** [review]
The diagrams are very nice! ❤️ I like how the Feeder, the starting point, sits almost in the upper-left corner and the Materialized, the finishing point, is located in the bottom-right corner, nicely flowing like a text on a page.

File: README.md (hunk 4 above)
Comment (nappex) [reply README.md:59]: agreed 🤝

File: README.md [hunk 5]
```diff
@@ -16,27 +16,47 @@ The flowchart below illustrates the flow of data between those parts:
...
+        spider_operator(🕷️ OPERATOR)
+        onigumo_downloader[DOWNLOADER]
     end
+
+    start([START]) --> onigumo_feeder[FEEDER]
+    onigumo_feeder -- .raw --> Crawling
+    onigumo_feeder -- .urls --> Crawling
+    onigumo_feeder -- .json --> Crawling
+    Crawling --> spider_materializer(🕷️ MATERIALIZER)
+    spider_materializer --> done([END])
```
Comment (Glutexo) [inline README.md:31]:
```suggestion
    start([START]) --> onigumo_feeder[FEEDER]
    
    onigumo_feeder -- .raw --> Crawling
    onigumo_feeder -- .urls --> Crawling
    onigumo_feeder -- .json --> Crawling
    
    Crawling --> spider_materializer(🕷️ MATERIALIZER)
    
    spider_materializer --> done([END])
```

I‘d split this into smaller groups.

File: README.md [hunk 6]
```diff
@@ -16,27 +16,47 @@ The flowchart below illustrates the flow of data between those parts:
...
+    end
+
+    onigumo_feeder -- .json --> spider_operator
+    spider_operator ---> spider_materializer
+    onigumo_feeder -- .urls --> onigumo_downloader
+    onigumo_feeder -- .raw --> spider_parser
+
+    spider_parser -. "<hash>.json" .-> spider_operator
+    onigumo_downloader -. "<hash>.raw" .-> spider_parser
+    spider_operator -. "<hash>.urls" .-> onigumo_downloader
```
Comment (Glutexo) [inline README.md:59]:
```suggestion
    onigumo_feeder -- .json --> spider_operator
    onigumo_feeder -- .urls --> onigumo_downloader
    onigumo_feeder -- .raw --> spider_parser

    spider_parser -. "<hash>.json" .-> spider_operator
    onigumo_downloader -. "<hash>.raw" .-> spider_parser
    spider_operator -. "<hash>.urls" .-> onigumo_downloader

    spider_operator ---> spider_materializer
```

See my [comment](https://github.com/Glutexo/onigumo/pull/237#discussion_r1720119236) from my previous review. I’d push the `spider_operator --→ spider_materializer` relation to the bottom, so the Spider subgraph arrows are coupled together

**Review by Glutexo — CHANGES_REQUESTED** [review]

File: README.md [hunk 7]
```diff
@@ -16,27 +16,51 @@ The flowchart below illustrates the flow of data between those parts:
...
+
+    Crawling --> spider_materializer(🕷️ MATERIALIZER)
+
+    spider_materializer --> done([END])
+
+    spider_operator -. "<hash>.urls" .-> onigumo_downloader
+    onigumo_downloader -. "<hash>.raw" .-> spider_parser
+    spider_parser -. "<hash>.json" .-> spider_operator
+```
+
```
Comment (Glutexo) [inline README.md:40]: We were thinking about adding some title to the second diagram. But I don’t know what it should say.

**Review by Glutexo — APPROVED** [review]
It may be worth updating the description “The flowchart below illustrates the flow of data between those parts:” to use a plural form “flowchart” now when there are two. But it’s not necessary: we may consider that to be one flowchart consisting of two parts. 😃

**Pull request merged**